docs/features/deconstruction.md

```diff
@@ -0,0 +1,180 @@
+```C#
+class C
+{
+    static void Main()
+    {
+        long x;
+        string y;
+
+        (x, y) = new C();
+        System.Console.WriteLine(x + "" "" + y);
+    }
+
+    public void Deconstruct(out int a, out string b)
+    {
+        a = 1;
+        b = ""hello"";
+    }
+}
+```
+
+Treat deconstruction of a tuple into existing variables as a kind of assignment, using the existing AssignmentExpression.
+
+
+###Deconstruction-assignment (deconstruction into into existing variables):
+This doesn't introduce any changes to the language grammar. We have an `assignment-expression` (also simply called `assignment` in the C# grammar) where the `unary-expression` (the left-hand-side) is a `tuple-literal`.
+In short, what this does is find a `Deconstruct` method on the expression on the right-hand-side of the assignment, invoke it, collect its `out` parameters and assign them to the variables on the left-hand-side.
+
+The existing assignment binding currently checks if the variable on its left-hand-side can be assigned to and if the two sides are compatible.
+It will be updated to support deconstruction-assignment, ie. when the left-hand-side is a tuple-literal/tuple-expression:
+
+- Each item on the left needs to be assignable and needs to be compatible with corresponding position on the right
+- Needs to handle nesting case such as `(x, (y, z)) = M();`, but note that the second item in the top-level group has no discernable type.
+
+The lowering for deconstruction-assignment would translate: `(expressionX, expressionY, expressionZ) = (expressionA, expressionB, expressionC)` into:
+
+```
+tempX = &evaluate expressionX
+tempY = &evaluate expressionY
+tempZ = &evaluate expressionZ
+
+tempRight = evaluate right and evaluate Deconstruct
+
+tempX = tempRight.A (including conversions)
+tempY = tempRight.B (including conversions)
+tempZ = tempRight.C (including conversions)
+
+“return/continue” with newTupleIncludingNames tempRight (so you can do get Item1 from the assignment)?
+```
+
+The evaluation order for nesting `(x, (y, z))` is:
+```
+tempX = &evaluate expressionX
+
+tempRight = evaluate right and evaluate Deconstruct
+
+tempX = tempRight.A (including conversions)
+tempLNested = tempRight.B (no conversions)
+
+tempY = &evaluate expressionY
+tempZ = &evaluate expressionZ
+
+tempRNest = evaluate Deconstruct on tempRight
+
+tempY = tempRNest.B (including conversions)
+tempZ = tempRNest.C (including conversions)
+
+```
+
+The evaluation order for the simplest cases (locals, fields, array indexers, or anything returning ref) without needing conversion:
+```
+evaluate side-effect on the left-hand-side variables
+evaluate Deconstruct passing the references directly in
+```
+
+Note that the feature is built around the `Deconstruct` mechanism for deconstructing types.
+`ValueTuple` and `System.Tuple` will rely on that same mechanism, except that the compiler may need to synthesize the proper `Deconstruct` methods.
+
```

Comment:
Is the intent to allow implicit calls to the Deconstruct method in source? If not then, perhaps it would be simpler to just specially recognize the types (without relying on the presence of the synthesized method) and let the lowering  to transfer element values.